### PR TITLE
Remove writing to SQLite in Experiment Tracking (Frontend)

### PR DESCRIPTION
**Pull request by jitu5**
## Description

Resolves [#1536](https://github.com/kedro-org/kedro-viz/issues/1536)

The edit run features of Experiment Tracking is now modified to store Run's metadata like `Bookmark`, `Notes` and `Title` in browser's `LocalStorage` Instead of writing to SQLite with `KedroViz-runs-metadata` as key.

```
KedroViz-runs-metadata: {
 <RUN_ID_1>: { bookmark: false, title: '', notes: '' },
 <RUN_ID_2>: { bookmark: false, title: '', notes: '' },
}  
```  

## Development notes

Run's metadata is stored in browser's `LocalStorage` using react-redux.

## Checklist

- [X] Read the [contributing](/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added new entries to the `RELEASE.md` file
- [X] Added tests to cover my changes

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -129,9 +129,9 @@
       }
     },
     "node_modules/@apollo/client": {
-      "version": "3.8.3",
-      "resolved": "https://registry.npmjs.org/@apollo/client/-/client-3.8.3.tgz",
-      "integrity": "sha512-mK86JM6hCpMEBGDgdO9U8ZYS8r9lPjXE1tVGpJMdSFUsIcXpmEfHUAbbFpPtYmxn8Qa7XsYy0dwDaDhpf4UUPw==",
+      "version": "3.8.5",
+      "resolved": "https://registry.npmjs.org/@apollo/client/-/client-3.8.5.tgz",
+      "integrity": "sha512-/ueWC3f1pFeH+tWbM1phz6pzUGGijyml6oQ+LKUcQzpXF6tVFPrb6oUIUQCbZpr6Xmv/dtNiFDohc39ra7Solg==",
```
Comment (ravi-kumar-pilla) [inline package-lock.json:134]: Are all these changes needed ?

File: src/actions/index.js [hunk 2]
```diff
@@ -273,3 +273,48 @@ export function toggleHoveredFocusMode(hoveredFocusMode) {
     hoveredFocusMode,
   };
 }
+
+export const TOGGLE_BOOKMARK = 'TOGGLE_BOOKMARK';
+
+/**
+ * Toggle bookmark of Run
+ * @param {boolean} bookmark
+ * @param {string} RunId
```
Comment (ravi-kumar-pilla) [inline src/actions/index.js:282]: [Nit] can we follow the camelCase for parameters

File: src/components/experiment-tracking/run-export-modal/run-export-modal.js [hunk 3]
```diff
@@ -15,13 +16,24 @@ const RunExportModal = ({
...
     useContext(ButtonTimeoutContext);
 
   const updateExportData = useCallback(() => {
-    setExportData(constructExportData(runMetadata, runTrackingData));
+    const mergedRunsMetadata = runMetadata?.map((run) => {
+      return {
+        ...run,
+        ...{
+          title: runsMetaData[run.id]?.title || run.id,
+          note: runsMetaData[run.id]?.title || '',
```
Comment (ravi-kumar-pilla) [inline src/components/experiment-tracking/run-export-modal/run-export-modal.js:31]: are the keys in the run metadata - **title** and **notes** . I am not sure if I understand this clearly. Is the **note** key corresponds to different metadata ?

File: src/components/experiment-tracking/run-metadata/run-metadata.js [hunk 4]
```diff
@@ -131,9 +146,9 @@ const RunMetadata = ({
                     <span
                       className="details-metadata__title-detail"
                       onClick={() => onTitleOrNoteClick(run.id)}
-                      title={sanitiseEmptyValue(run.title)}
+                      title={sanitiseEmptyValue(getTitleByRunId(run.id))}
```
Comment (ravi-kumar-pilla) [inline src/components/experiment-tracking/run-metadata/run-metadata.js:149]: can we have runTitle and runNotes initialized and set a value before returning the html template ? This way we can avoid the function call multiple times in the template.

**Review by ravi-kumar-pilla — COMMENTED** [review]
Great work @jitu5 !! 

I left few comments and suggestions. I will try to run the branch locally and test the functionality after this. Thank you !

File: package-lock.json (hunk 1 above)
Comment (jitu5) [reply package-lock.json]: No, I updated it.

File: src/actions/index.js (hunk 2 above)
Comment (jitu5) [reply src/actions/index.js]: Done.

File: src/components/experiment-tracking/run-export-modal/run-export-modal.js (hunk 3 above)
Comment (jitu5) [reply src/components/experiment-tracking/run-export-modal/run-export-modal.js]: Corrected, it was typo. Thanks

File: src/components/experiment-tracking/run-metadata/run-metadata.js (hunk 4 above)
Comment (jitu5) [reply src/components/experiment-tracking/run-metadata/run-metadata.js]: Done.

**Review by vladimir-mck — APPROVED** [review]

File: src/components/experiment-tracking/experiment-primary-toolbar/experiment-primary-toolbar.js [hunk 5]
```diff
@@ -26,14 +27,13 @@ export const ExperimentPrimaryToolbar = ({
   showRunDetailsModal,
   sidebarVisible,
   setShowRunExportModal,
+  onToggleBookmark,
+  runsMetaData,
```
Comment (tynandebold) [inline src/components/experiment-tracking/experiment-primary-toolbar/experiment-primary-toolbar.js:31]: The `d` in `MetaData` should be lowercase.

```suggestion
  runsMetadata,
```

It should be like that in all other places, too.

File: src/components/experiment-tracking/run-metadata/run-metadata.js [hunk 6]
```diff
@@ -102,6 +103,27 @@ const RunMetadata = ({
     setShowRunDetailsModal(true);
   };
 
+  const getNotesByRunId = (runId) => {
+    if (runsMetaData[runId]) {
+      return runsMetaData[runId][RUN_NOTES] || '';
+    }
+    return '';
```
Comment (tynandebold) [inline src/components/experiment-tracking/run-metadata/run-metadata.js:110]: Empty line before `return` statements.
```suggestion
    if (runsMetaData[runId]) {
      return runsMetaData[runId][RUN_NOTES] || '';
    }
    
    return '';
```

File: src/components/experiment-tracking/run-metadata/run-metadata.js [hunk 7]
```diff
@@ -102,6 +103,27 @@ const RunMetadata = ({
...
+  };
+
+  const getTitleByRunId = (runId) => {
+    if (runsMetaData[runId]) {
+      return runsMetaData[runId][RUN_TITLE] || runId;
+    }
+    return runId;
```
Comment (tynandebold) [inline src/components/experiment-tracking/run-metadata/run-metadata.js:117]:
```suggestion
    if (runsMetaData[runId]) {
      return runsMetaData[runId][RUN_TITLE] || runId;
    }
    
    return runId;
```

**Review by tynandebold — COMMENTED** [review]
Looks great and works very well. Thank you for this.

I've left a couple of small comments about spelling and code style. There's one other things too, and that's whenever there's a run that's bookmarked, that first bookmarked run should be selected when the page is loaded. Right now, that's not happening. Please double check that, and when fixed, it'll be good to go!

File: src/components/experiment-tracking/experiment-primary-toolbar/experiment-primary-toolbar.js (hunk 5 above)
Comment (jitu5) [reply src/components/experiment-tracking/experiment-primary-toolbar/experiment-primary-toolbar.js]: Done.

File: src/components/experiment-tracking/run-metadata/run-metadata.js (hunk 6 above)
Comment (jitu5) [reply src/components/experiment-tracking/run-metadata/run-metadata.js]: done.

File: src/components/experiment-tracking/run-metadata/run-metadata.js (hunk 7 above)
Comment (jitu5) [reply src/components/experiment-tracking/run-metadata/run-metadata.js]: done.

**Comment by jitu5** [comment]
@tynandebold Thanks for catching this, now fixed. 
> whenever there's a run that's bookmarked, that first bookmarked run should be selected when the page is loaded.

**Review by tynandebold — APPROVED** [review]
Looks good!

**Comment by ravi-kumar-pilla** [comment]
Hi @jitu5 , I tried running it locally, the mutations are saved to localStorage as expected. Though while editing the run, I found a minor issue -
1. Edit run details.
2. Click on Cancel
3. Reopen the Edit run modal
4. Expected (I guess) - The modal should display the actual run details. But, the modal displays the discarded changes as below -

sqlite_mutation_issue

**Comment by tynandebold** [comment]
> Hi @jitu5 , I tried running it locally, the mutations are saved to localStorage as expected. Though while editing the run, I found a minor issue -
> 
> 1. Edit run details.
> 2. Click on Cancel
> 3. Reopen the Edit run modal
> 4. Expected (I guess) - The modal should display the actual run details. But, the modal displays the discarded changes as below -
> 
> sqlite_mutation_issue [ sqlite_mutation_issue ](https://user-images.githubusercontent.com/87735534/276388192-a36d868a-d4d3-4e3c-9c7b-544670a0e1e5.gif) [ ](https://user-images.githubusercontent.com/87735534/276388192-a36d868a-d4d3-4e3c-9c7b-544670a0e1e5.gif)

Good catch! This doesn't happen in the demo site, so it's something to do with this PR.

**Comment by jitu5** [comment]
@ravi-kumar-pilla @tynandebold 
Thanks Ravi for catching this. It was not reseting values on modal close, now fixed.

**Review by ravi-kumar-pilla — APPROVED** [review]
Awesome work !!

**Pull request merged**